emu/src/ppu.rs

```diff
@@ -1,2 +1,130 @@
 pub const DISPLAY_WIDTH: usize = 240;
 pub const DISPLAY_HEIGHT: usize = 160;
+pub const PALETTE_SIZE: usize = 16;
+
+pub const BG_PALETTE_ADDRESS: u32 = 0x05000000;
+pub const OBJ_PALETTE_ADDRESS: u32 = 0x05000200;
+
+use std::{cell::RefCell, rc::Rc};
+
+#[cfg(feature = "debug")]
+use rand::Rng;
+
+use crate::{
+    internal_memory::InternalMemory,
+    render::palette_color::{PaletteColor, PaletteType},
+};
+
+pub struct PixelProcessUnit {
+    internal_memory: Rc<RefCell<InternalMemory>>,
+}
+
+impl PixelProcessUnit {
+    pub fn new(internal_memory: Rc<RefCell<InternalMemory>>) -> Self {
+        Self { internal_memory }
+    }
+
+    fn get_array_color(&self, index: usize, palette_type: PaletteType) -> [u8; 2] {
+        match palette_type {
+            PaletteType::BG => [
+                self.internal_memory.borrow().bg_palette_ram[index],
+                self.internal_memory.borrow().bg_palette_ram[index + 1],
+            ],
+            PaletteType::OBJ => [
+                self.internal_memory.borrow().obj_palette_ram[index],
+                self.internal_memory.borrow().obj_palette_ram[index + 1],
+            ],
+            PaletteType::FULL => match index {
+                0..=255 => self.get_array_color(index * 2, PaletteType::BG),
+                256..=551 => self.get_array_color((index - 256) * 2, PaletteType::OBJ),
```

Comment:
Why `551`? We have 512 colors in total, right?

Reply:
Yes, my bad :)